x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -0,0 +1,740 @@
+        if (authenticationToken instanceof JwtAuthenticationToken jwtAuthenticationToken) {
+            LOGGER.trace("Realm [{}] supports JwtAuthenticationToken.", super.name());
+            final JWSHeader jwsHeader = jwtAuthenticationToken.getJwsHeader();
+            final JWTClaimsSet jwtClaimsSet = jwtAuthenticationToken.getJwtClaimsSet();
+            final String clientAuthorizationSharedSecret = jwtAuthenticationToken.getClientAuthorizationSharedSecret().toString();
+
+            // Filter steps (before any validation)
+
+            // 1. Skip JWT if signature algorithm does not match any of the signature algorithms allowed by this realm.
+            final JWSAlgorithm jwsSignatureAlgorithm = jwsHeader.getAlgorithm();
+            if ((jwsSignatureAlgorithm == null) || (this.allowedSignatureAlgorithms.contains(jwsSignatureAlgorithm.getName()) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow signature algorithm [%s]. Allowed signature algorithms are %s.",
+                    super.name(),
+                    jwsSignatureAlgorithm,
+                    this.allowedSignatureAlgorithms
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug(
+                "Realm [%s] allows signature algorithm [%s]. Allowed signature algorithms are %s.",
+                super.name(),
+                jwsSignatureAlgorithm,
+                this.allowedSignatureAlgorithms
+            );
+
+            // 2. Skip JWT if issuer does not match the issuer allowed by this realm.
+            final String jwtIssuer = jwtClaimsSet.getIssuer();
+            if ((jwtIssuer == null) || (this.allowedIssuer.contains(jwtIssuer) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow issuer [%s]. Allowed issuer is [%s].",
+                    super.name(),
+                    jwtIssuer,
+                    this.allowedIssuer
+                );
+                LOGGER.debug(msg);
+                listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                return;
+            }
+            LOGGER.debug("Realm [%s] allows issuer [%s]. Allowed issuer is [%s].", super.name(), jwtIssuer, this.allowedIssuer);
+
+            // 3. Skip JWT if audience does not match any of the audiences allowed by this realm.
+            final List<String> jwtAudiences = jwtClaimsSet.getAudience();
+            if ((jwtAudiences == null) || (this.allowedAudiences.stream().anyMatch(jwtAudiences::contains) == false)) {
+                final String msg = String.format(
+                    Locale.ROOT,
+                    "Realm [%s] does not allow audiences %s. Allowed audiences are %s.",
```

Comment:
```suggestion
                    "Realm [%s] does not allow audiences [%s]. Allowed audiences are %s.",
```

Reply:
JWT aud claim is a list, not a String. I need to wrap args 2 and 3 with String.join(), and include square brackets for both.

https://openid.net/specs/openid-connect-core-1_0.html#IDToken
> aud
REQUIRED. Audience(s) that this ID Token is intended for. It MUST contain the OAuth 2.0 client_id of the Relying Party as an audience value. It MAY also contain identifiers for other audiences. In the general case, the aud value is an array of case sensitive strings. In the common special case when there is one audience, the aud value MAY be a single case sensitive string.